chat-client.py

```diff
@@ -0,0 +1,64 @@
+import socket
+import selectors
+import errno
+import sys
+import time
+
+HEADER_LENGTH = 10
+HOST = '127.0.0.1'
+PORT = 54321
+
+print('\nWelcome to the chat server!\n')
+time.sleep(1)
+print('Hit enter to send message or refresh message feed\n')
+time.sleep(1)
+print('Type "quit" to leave the chat\n')
+time.sleep(1)
+my_username = input('Enter username: ')
+
+# Set up client socket
+sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+sock.connect((HOST, PORT))
+sock.setblocking(False)
+
+# Send username to server with a header
+my_username_enc = my_username.encode("utf-8")
+my_username_header = f'{len(my_username_enc):<{HEADER_LENGTH}}'.encode("utf-8")
+sock.send(my_username_header + my_username_enc)
+
+while True:
+	msg = input(f'{my_username}> ')
+	if msg.lower() == 'quit':
+		sock.close()
+		print('Connection closed')
+		sys.exit()
+	elif msg:
+		# Send message
+		msg_enc = msg.encode("utf-8")
+		msg_header = f'{len(msg_enc):<{HEADER_LENGTH}}'.encode("utf-8")
+		sock.send(msg_header + msg_enc)
+	try:
+		# Receive messages
+		while True:
+			user_header = sock.recv(HEADER_LENGTH)
```

Comment:
You should check the length of user_header because `sock.recv()` will read *up to* HEADER_LENGTH

Reply:
Same goes for the other reads. 

Tip: Always carefully validate data received from an external source, even if it seems tedious

Reply:
Again I have introduced a function `recv_msg()` that checks all bytes are received and replaced all calls to `recv()` with it. Some slightly strange behaviour though: you'll notice in the function definition in chat-client.py the `try: except:` block is commented out. This is because when the user_header was received (first call to recv_msg()) the exception was getting triggered. However just running the code inside the try: block it works fine. It also works in chat-server.py WITH the `try: except:` included. Any idea whats going on here?